specification/under_development/Base-Profile.md

```diff
@@ -0,0 +1,160 @@
+|---------------------------|-------------------|-------------|
+| call    | Used within a function block to invoke any one of the declared qis functions and the output recording functions. | (none) |
+| ret | ... | ... |
+| inttoptr | ... | ... |
+
+## Data Types and Values
+
+Within the base profile, defining local variables is not supported. Arguments to calls correspondingly are expected to be constant values. 
+
+Qubits and results are passes as a pointer of type `%Qubit*` and `%Result*` respectively, where the pointer itself rather than the memory location identifies the qubit or result: For the purpose of passing them as arguments in function calls, a 64-bit signed integer value is cast to the appropriate pointer type. The signed integer that is case must be in the interval `[0, nrQubits)` for `%Qubit*` and `[0, nrResults)` for `%Result*`, where `nrQubits` and `nrResults` are the required number of qubits and results defined by entry point attributes.
+
+Only the `%Qubit` and `%Result` data types are required to be supported by all backends. 
+
+Integers and double-precision floating point numbers are available as in full QIR;
+however, computations using these numeric types are not available.
+
+### Qubit and Result Usage
+
+Qubits may not be used after they have been measured. 
+Qubits and results need to be numbered consecutively. 
+No qpu support needed for measuring individual qubits, but backend would support output selection?
+
+## Quantum Instruction Set
+
+The QIR specification contains a [table]() of commonly used instructions along with their signatures and a description of their functionality.
+Backends a **not** required to support all of these. Instead, each backend will declare which of these instructions it supports. Only native instructions on the quantum processor should be listed. A wider set of instructions can be made available in the form of a library that defines set qir functions calling into the natively supported instruction set. This library is provided in the form of a bitcode file and pulled in as part of the QIR compilation stage. 
+
+## Output Recording
+
+The following functions are declared and used to record the program output: 
+| Function                  | Signature         | Description |
+|---------------------------|-------------------|-------------|
+| __quantum__rt__tuple_start_record_output    | `void()`  | Inserts a marker in the output log that indicates the start of a tuple. |
+| __quantum__rt__tuple_end_record_output    | `void()`  | Inserts a marker in the output log that indicates the end of a tuple. |
+| __quantum__rt__array_start_record_output    | `void()`  | Inserts a marker in the output log that indicates the start of an array. |
+| __quantum__rt__array_end_record_output    | `void()`  | Inserts a marker in the output log that indicates the end of an array. |
+| __quantum__rt__result_record_output   | `void(%Result*)`  | Adds a measurement result to the output log. |
```

Comment:
> I feel it would be much more intuitive and useful for further profiles if the entrypoint returned a tuple (or list, whichever is easiest to represent in LLVM) with a length of the classical bits returned, and the various results put in there instead.

Don't the `__quantum__rt__*` record output functions offer a different utility though? My understanding here is that these could also be interleaved throughout different parts of the circuit, not just at the end. Maybe we're demonstrating them as a means to output final results now, but I can see these being useful for mid-circuit measurement recording and capturing various results as a program is executed. By no means am I certain of this use case though!

Reply:
> Don't the __quantum__rt__* record output functions offer a different utility though? My understanding here is that these could also be interleaved throughout different parts of the circuit, not just at the end.

This is in part why I'm concerned about them. 

Just a few thoughts:

1. We already have well-known ways for defining lists, tuples and returns - both in the general and LLVM sense (afaik). What's the benefit of creating new ways to do so? This excludes the complication of making sure start/end always match up, nested method calls doing fun things, etc.
2. Conceptually you're recording results for the entire QIR file, not the results from a method that just happens to be an entry point. This is OK when you've only got one entry and exit point, far less-so when you have more than that.

Lets give a practical example of that last one: lets say you have 2 entry-point methods that can be called in a QIR file and they do whatever they're meant to do by calling each other. Lets just call an arbitrary EP, and then it calls the other to do _something_ quantum, which returns that result to our called EP, which then does something else quantum based off these results and then returns. Because both are valid entry-points into the IR they both have record intrinsics at the end of each method, so will incorrectly return 2 values, or you have to start adding conditions about when something gets recorded and this complexity spirals. (While, yes, EP's in this example need to return results it's both an easy example for a universal issue and it might happen at some point anyway)

This issue gets more amusing the more higher-level concepts you add in like threads, exceptions, the above example having shared state between runs, etc.

I expect the reason the intrinsics exist the way they are is for the runtime to hook into and so you can combine a QIR walker, runtime and backend in relative isolation, which isn't a bad thing overall. Would love to know the exact reasons though.

Reply:
Great discussion point - let's dive into this in the next workstream meeting.

Reply:
Just for posterity: my concerns won't be a problem due to the entry-point methods effectively being auto-generated wrappers that only have the recording intrinsics in and nothing else (beyond a potential program return value), and are never called from elsewhere in the program being purely an entry and exit point for commandline/isolated running. Also the record intrinsics don't appear outside these methods, so the cross-cutting won't happen.